### PR TITLE
P3 Terminal Velocity

### PR DESCRIPTION
**Pull request by anastasia-popova**
## Purpose 

Implement Terminal Velocity in the P3 Scheme, compare to the plots in the Mildbrant and Morrison (2015) paper 

## To-do

- [x]  transfer to integrate function() 
- [x] move mass function back to the plots 
- [x] implement the Chen 2022 velocity 
- [x] add tests for velocity
- [x] create plots comparable to the P3 paper 
- [ ] compare plots

## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: docs/src/plots/P3SchemePlots.jl [hunk 1]
```diff
@@ -9,6 +9,66 @@ FT = Float64
...
+        return mass_nl(p3, D)             # dense nonspherical ice
+    end
+    if th.D_cr > D >= th.D_gr
+        return mass_s(D, th.ρ_g)          # graupel
+    end
+    if D >= th.D_cr
+        return mass_r(p3, D, F_r)         # partially rimed ice
+    end
+
+    # TODO - would something like this be better?
```
Comment (trontrytel) [inline docs/src/plots/P3SchemePlots.jl:63]: We can delete this comment.If it's no longer part of source code we don't need to optimize

File: src/P3Scheme.jl [hunk 2]
```diff
@@ -288,6 +229,51 @@ function DSD_N₀(p3::PSP3, N::FT, λ::FT) where {FT}
...
+        1 - small, spherical ice 
+        2 - large, unrimed ice 
+        3 - dense, nonspherical ice 
+        4 - graupel 
+        5 - partially rimed ice 
+        6 - second half of partially rimed ice (only for a)
+"""
+function get_coeffs(p3::PSP3, th, F_r::FT) where{FT}
+    α_va = α_va_si(p3)
+    m = [π / 6 * p3.ρ_i, α_va, α_va, π / 6 * th.ρ_g, α_va / (1- F_r)]
```
Comment (trontrytel) [inline src/P3Scheme.jl:270]:
```suggestion
    m = [FT(π) / 6 * p3.ρ_i, α_va, α_va, π / 6 * th.ρ_g, α_va / (1- F_r)]
```

File: Project.toml [hunk 3]
```diff
@@ -29,7 +30,8 @@ DocStringExtensions = "0.8, 0.9"
 EnsembleKalmanProcesses = "1.1.5"
 ForwardDiff = "0.10"
 MLJ = "0.20"
+QuadGK = "2.9.4"
 RootSolvers = "0.3, 0.4"
 SpecialFunctions = "1, 2"
 Thermodynamics = "0.12.4"
-julia = "1.6"
+julia = "1.6"
```
Comment (nefrathenrici) [inline Project.toml:37]:
```suggestion
julia = "1.6"

```

File: docs/src/plots/P3SchemePlots.jl [hunk 4]
```diff
@@ -9,6 +9,58 @@ FT = Float64
...
+    end
+    if th.D_gr > D >= D_th
+        return mass_nl(p3, D)             # dense nonspherical ice
+    end
+    if th.D_cr > D >= th.D_gr
+        return mass_s(D, th.ρ_g)          # graupel
+    end
+    if D >= th.D_cr
+        return mass_r(p3, D, F_r)         # partially rimed ice
+    end
```
Comment (nefrathenrici) [inline docs/src/plots/P3SchemePlots.jl:61]: Should these use `elseif`?

File: src/P3Scheme.jl [hunk 5]
```diff
@@ -288,6 +231,26 @@ function DSD_N₀(p3::PSP3, N::FT, λ::FT) where {FT}
...
+    integrate(a, b, c1, c2, c3)
+
+ - a - lower bound 
+ - b - upper bound 
+ - c1, c2, c3 - respective constants 
+
+ Integrates the function c1 * D ^ (c2) * exp(-c3 * D) dD from a to b 
+    Returns the result
+"""
+function integrate(a::FT, b::FT, c1::FT, c2::FT, c3::FT) where {FT}
```
Comment (nefrathenrici) [inline src/P3Scheme.jl:244]: This function should have a more descriptive name, since it integrates a specific function.

File: src/P3Scheme.jl [hunk 6]
```diff
@@ -451,10 +409,500 @@ function distribution_parameter_solver(
...
+                            (F_r * π / 4 * D^2 + (1 - F_r) * p3.γ * D^p3.σ)^3 /
+                            (9 * π * (α_va / (1 - F_r) * D^p3.β_va)^2)
+                        )^κ *
+                        ai[i] *
+                        D^(bi[i]) *
+                        exp(-ci[i] * D) *
+                        (α_va / (1 - F_r) * D^p3.β_va) *
+                        N_0 *
+                        D^μ *
+                        exp(-λ * D),
```
Comment (nefrathenrici) [inline src/P3Scheme.jl:604]: This anonymous function could be saved to a variable since it repeats 2 more times.

File: src/P3Scheme.jl [hunk 7]
```diff
@@ -451,10 +409,500 @@ function distribution_parameter_solver(
...
+                            p3.ρ_i^2 *
+                            (F_r * π / 4 * D^2 + (1 - F_r) * p3.γ * D^p3.σ)^3 /
+                            (9 * π * (α_va / (1 - F_r) * D^p3.β_va)^2)
+                        )^κ *
+                        ai[i] *
+                        D^(bi[i]) *
+                        exp(-ci[i] * D) *
+                        N_0 *
+                        D^μ *
+                        exp(-λ * D),
```
Comment (nefrathenrici) [inline src/P3Scheme.jl:812]: Same as my comment above, this function could be saved since it is used 3 times.

**Review by nefrathenrici — COMMENTED** [review]
Hi @anastasia-popova ! I just left a few comments that I hope are helpful

File: src/P3Scheme.jl [hunk 8]
```diff
@@ -288,6 +231,26 @@ function DSD_N₀(p3::PSP3, N::FT, λ::FT) where {FT}
     return N / Γ(1 + μ) * λ^(1 + μ)
 end
 
+"""
+    integrate_to_gamma(a, b, c1, c2, c3)
```
Comment (trontrytel) [inline src/P3Scheme.jl:235]: This is much neater and more readable. Thank you!

File: Project.toml [hunk 9]
```diff
@@ -29,6 +30,7 @@ DocStringExtensions = "0.8, 0.9"
 EnsembleKalmanProcesses = "1.1.5"
 ForwardDiff = "0.10"
 MLJ = "0.20"
+QuadGK = "2.9.4"
```
Comment (nefrathenrici) [inline Project.toml:33]: Does the compat need to be so specific? Would 2.9 work?

File: src/P3Scheme.jl [hunk 10]
```diff
@@ -451,10 +415,447 @@ function distribution_parameter_solver(
...
-            10,
+            5,
         ).root
 
     return (; λ = exp(x), N_0 = DSD_N₀(p3, N, exp(x)))
 end
 
+"""
+    terminal_velocity_mass(p3, Chen2022, q, N, ρ_r, F_r)
+
```
Comment (trontrytel) [inline src/P3Scheme.jl:426]: There is a lot of repetition here. I'll try to combine those two functions into one

File: src/P3Scheme.jl [hunk 11]
```diff
@@ -451,10 +415,447 @@ function distribution_parameter_solver(
...
+        if F_r == 0
+            # Velocity coefficients for small particles
+            (ai, bi, ci) = CO.Chen2022_vel_coeffs_small(Chen2022, ρ_a)
+            v += integrate_to_gamma(
+                FT(0),
+                D_th,
+                FT(π) / 6 * p3.ρ_i * ai[i] * N_0,
+                bi[i] + μ + 3,
+                ci[i] + λ,
+            )
```
Comment (nefrathenrici) [inline src/P3Scheme.jl:474]: These lines are run in both cases, so they could be hoisted out of the ifelse

**Review by nefrathenrici — COMMENTED** [review]
Looks good! I agree with @trontrytel, there is some duplicate code that could be merged.

**Comment by trontrytel** [comment]
Done via https://github.com/CliMA/CloudMicrophysics.jl/pull/370

**Pull request closed without merging**